mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp

```diff
@@ -0,0 +1,218 @@
+/// 2. Load the object
+/// 3. Initialize the dimensions of the object with the expected results using
+/// the operation with name \p opName from the GPU dialect
+void convertToFullObject(ConversionPatternRewriter &rewriter, StringRef opName,
+                         Operation *op, int64_t dimensions) {
+  // This conversion is platform dependent
+  assert(dimensions <= gpu::GPUDialect::getNumWorkgroupDimensions() &&
+         "Invalid number of dimensions");
+  const auto loc = op->getLoc();
+  const auto targetIndexTy = rewriter.getIntegerType(64);
+  const auto getIndexTy = rewriter.getIntegerType(32);
+  const auto dimMtTy = MemRefType::get(dimensions, targetIndexTy, {}, 4);
+  // Allocate
+  const auto resTy = op->getResultTypes()[0];
+  const auto alloca = static_cast<Value>(
+      rewriter.create<memref::AllocaOp>(loc, MemRefType::get(1, resTy)));
+  // Load
+  const auto zero =
+      static_cast<Value>(rewriter.create<arith::ConstantIndexOp>(loc, 0));
+  const auto res = static_cast<Value>(
+      rewriter.replaceOpWithNewOp<memref::LoadOp>(op, alloca, zero));
+  const auto argumentTypes =
+      rewriter.getTypeArrayAttr({MemRefType::get(1, resTy, {}, 4), getIndexTy});
+  const auto functionName = rewriter.getAttr<FlatSymbolRefAttr>("operator[]");
+  // Initialize
+  for (int64_t i = 0; i < dimensions; ++i) {
+    const auto index = static_cast<Value>(
+        rewriter.create<arith::ConstantIntOp>(loc, i, getIndexTy));
+    const auto val = static_cast<Value>(rewriter.create<arith::IndexCastOp>(
+        loc, targetIndexTy, getDimension(rewriter, loc, opName, i)));
+    const auto ptr = createGetOp(rewriter, loc, dimMtTy, res, index,
+                                 argumentTypes, functionName);
+    rewriter.create<memref::StoreOp>(loc, val, ptr, zero);
+  }
+}
+
+/// Converts n-dimensional operations to operations of name \p opName, from the
+/// GPU dialect.
+///
+/// There are three possible cases here:
+/// 1. No argument is passed (see convertToFullObject())
+/// 2. A constant argument is passed (see converWithConstIndex())
+/// 3. A non-constant argument is passed (see convertWithVarIndex())
+void rewrite(StringRef opName, Operation *op, ValueRange operands,
+             ConversionPatternRewriter &rewriter) {
+  switch (op->getNumOperands()) {
+  case 0: {
+    const auto dimensions = getDimensions(op->getResultTypes()[0]);
+    convertToFullObject(rewriter, opName, op, dimensions);
+    break;
+  }
+  case 1: {
+    const auto dimension = operands[0];
+    if (const auto definingOp = dimension.getDefiningOp<arith::ConstantOp>()) {
+      convertWithConstIndex(rewriter, opName, op, definingOp);
+    } else {
+      // TODO: Do not rely in a default number of dimensions; take into
+      // account kernel dimensionality. This should be available in the parent
+      // function.
+      constexpr int64_t defaultDimensions{3};
+      convertWithVarIndex(rewriter, opName, op, defaultDimensions, dimension);
```

Comment:
Otherwise, we have to create an array, store all the values (3) and load with `index` offset to mimic the original semantics. This is similar to the LLVM code DPC++ generates in these cases.